kernel/src/schema/mod.rs

```diff
@@ -212,33 +216,107 @@ impl StructField {
             .collect()
     }
 
-    /// Applies physical name mappings to this field
+    /// Applies physical name mappings to this field. If the `column_mapping_mode` is
+    /// [`ColumnMappingMode::Id`], then each StructField will have its parquet field id in the
+    /// [`ColumnMetadataKey::ParquetFieldId`] metadata field.
+    ///
     ///
     /// NOTE: Caller affirms that the schema was already validated by
     /// [`crate::table_features::validate_schema_column_mapping`], to ensure that annotations are
     /// always and only present when column mapping mode is enabled.
-    pub fn make_physical(&self) -> Self {
-        struct MakePhysical;
+    pub fn make_physical(&self, column_mapping_mode: ColumnMappingMode) -> Self {
+        struct MakePhysical {
+            column_mapping_mode: ColumnMappingMode,
+        }
         impl<'a> SchemaTransform<'a> for MakePhysical {
             fn transform_struct_field(
                 &mut self,
                 field: &'a StructField,
             ) -> Option<Cow<'a, StructField>> {
                 let field = self.recurse_into_struct_field(field)?;
-                Some(Cow::Owned(field.with_name(field.physical_name())))
+
+                let metadata = field.logical_to_physical_metadata(self.column_mapping_mode);
+                let name = match self.column_mapping_mode {
+                    ColumnMappingMode::None => field.name().to_owned(),
+                    ColumnMappingMode::Id | ColumnMappingMode::Name => {
+                        // Assert that the physical name is present
+                        debug_assert!(field
+                            .metadata
+                            .get(ColumnMetadataKey::ColumnMappingPhysicalName.as_ref())
+                            .is_some_and(|x| matches!(x, MetadataValue::String(_))));
+                        field.physical_name().to_owned()
+                    }
+                };
+
+                Some(Cow::Owned(StructField {
+                    name,
+                    data_type: field.data_type.clone(),
+                    nullable: field.nullable,
+                    metadata,
+                }))
             }
         }
         // NOTE: unwrap is safe because the transformer is incapable of returning None
-        MakePhysical
-            .transform_struct_field(self)
-            .unwrap()
-            .into_owned()
+        MakePhysical {
+            column_mapping_mode,
+        }
```

Comment:
nit: maybe make into tuple or add constructor for simpler usage/fewer LOC? (though if we add constructor might not be worth LOC?)

Reply:
Yeah constructor adds more LOC. Tuple works, but then we match on `self.0`. I decided this was the most explicit.